dashboard/cron.py

```diff
@@ -54,19 +54,39 @@ def util_function(UDW_course_id, sql_string, mysql_table, table_identifier=None)
     # write to MySQL
     try:
         df.to_sql(con=engine, name=mysql_table, if_exists='append', index=False)
-    except Exception as e:
+    except Exception:
         logger.exception(f"Error running to_sql on table {mysql_table}")
         raise
 
     # returns the row size of dataframe
     return f"inserted + {str(df.shape[0])} rows in table {mysql_table} for course {UDW_course_id}\n"
 
+# Exit with log if UDW dataset is missing courses
+def data_validation (incoming_courses,existing_courses):
+    incoming_courses = df['course_id']
+    existing_courses = get_queryset().get_supported_courses()
```

Comment:
we have discussed yesterday that the get_supported_courses() function merely returns what's inside the Admin configuration. 

It does not verify against the UDW to check whether the given course id is a valid Canvas course id.

The whole purpose of this request is to check the course id against UDW.

Reply:
That is accurate.
The existing courses are pulled from the Admin config.

The feed from UDW is "incoming_courses"

Reply:
This is new function "data_validation" added in this pull request. 

Where is it used?

How do you know the parameter "incoming_courses" is the feed from UDW?

Reply:
This is new function "data_validation" added in this pull request.

Where is it used?

I created it as a function so it can be called in the future.
I want to reuse this code so that we can load individual courses when the data feed is incomplete.
It is currently taking in incoming and existing courses as arguments and comparing them to see if data is missing.
If data is missing, the cron fails.

Reply:
How do you know the parameter "incoming_courses" is the feed from UDW?

There is a pandas dataframe generated earlier in the code.
Would you like me to review that with you?

Reply:
I will go ahead and take initiative here.

The code itself is very well documented, so it will be easy to review on how I know it is the feed.
```

# the util function
def util_function(UDW_course_id, sql_string, mysql_table, table_identifier=None):
    df = pd.read_sql(sql_string, conns['UDW'])
    logger.debug(df)

    # Sql returns boolean value so grouping course info along with it so that this could be stored in the DB table.
    if table_identifier == 'weight' and UDW_course_id:
        df['course_id']=UDW_course_id
        df.columns=['consider_weight','course_id']

    # drop duplicates
    df.drop_duplicates(keep='first', inplace=True)

    logger.debug(" table: " + mysql_table + " insert size: " + str(df.shape[0]))

    # write to MySQL
    try:
        df.to_sql(con=engine, name=mysql_table, if_exists='append', index=False)
    except Exception:
        logger.exception(f"Error running to_sql on table {mysql_table}")
        raise

    # returns the row size of dataframe
    return f"inserted + {str(df.shape[0])} rows in table {mysql_table} for course {UDW_course_id}\n"
```

Lines 39-62 of cron.py

Reply:
I then extract the course id's from the UDW dataframe with 
`incoming_courses = df['course_id']`

Does that clarify your inquiry?

Reply:
>I then extract the course id's from the UDW dataframe with
>incoming_courses = df['course_id']
> Does that clarify your inquiry?

No, it does not. 

The df variable is not a global variable, nor it is defined within data_validation().

This line will throw error when the data_validation() is called.

Reply:
Thanks for the insight.
That hasn't been my experience, but I will investigate it.
Have you tested it locally to confirm an error?

Reply:
Updated for 1:1